addons/mass_mailing/models/mass_mailing.py

```diff
@@ -629,19 +686,14 @@ def _get_opt_out_list(self):
         self.ensure_one()
         opt_out = {}
         target = self.env[self.mailing_model_real]
-        mail_field = 'email' if 'email' in target._fields else 'email_from'
-        if 'opt_out' in target._fields:
-            # avoid loading a large number of records in memory
-            # + use a basic heuristic for extracting emails
-            query = """
-                SELECT lower(substring(%(mail_field)s, '([^ ,;<@]+@[^> ,;]+)'))
-                  FROM %(target)s
-                 WHERE opt_out AND
-                       substring(%(mail_field)s, '([^ ,;<@]+@[^> ,;]+)') IS NOT NULL;
-            """
-            query = query % {'target': target._table, 'mail_field': mail_field}
-            self._cr.execute(query)
-            opt_out = set(m[0] for m in self._cr.fetchall())
+        if self.mailing_model_real == "mail.mass_mailing.contact":
+            # if user is opt_out on One list but not on another, send the mail anyway
+            target_list_contacts = self.env['mail.mass_mailing.list_contact_rel'].search(
+                [('list_id', 'in', self.contact_list_ids.ids)])
+            opt_out_contacts = target_list_contacts.filtered(lambda rel: rel.opt_out).mapped('contact_id.email')
+            opt_in_contacts = target_list_contacts.filtered(lambda rel: not rel.opt_out).mapped('contact_id.email')
+            opt_out = set(c for c in opt_out_contacts if c not in opt_in_contacts)
```

Comment:
I don't understand that code. I suppose it comes from having several mailing lists, sometimes opt-out, sometimes not and a contact can be in both state. What is the expected behavior (by the way it could be documented somewhere) ? If i understand well you fetch contact from the mailing lists that have at least one "not opt-outed" list.

Could we try to have an optimized SQL query ? Search on contacts with a join on contact_rel, having at least one not opt-outed contact_rel meaning you are considered as opt-in.